Improve kube-proxy performance by tackling down TODO in  proxysocket.go

roackb2 opened the pull request:
At `/pkg/proxy/userspace/proxysocket.go`, in the `tryConnect` function, there's a TODO:

``` go
// TODO: This could spin up a new goroutine to make the outbound connection,
// and keep accepting inbound traffic.
```

 In the function `ProxyLoop`, I extract the part of tryConnect to another function and run it as a goroutine,
makes the `ProxyLoop` able to accept next inbound request immediately instead of waiting the outbound connection to finish.

roackb2 committed:
tackle down TODO in proxysocket.go
move tryConnect to another function and run it as goroutine

roackb2 commented:
I signed it!

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -133,15 +133,18 @@ func (tcp *tcpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
-		}
-		// Spin up an async copy loop.
-		go proxyTCP(inConn.(*net.TCPConn), outConn.(*net.TCPConn))
+		go connecBalancer(service, inConn, proxier)
```
s/connecBalancer/connectToBackend/

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -133,15 +133,18 @@ func (tcp *tcpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
+	outConn, err := tryConnect(service, inConn.(*net.TCPConn).RemoteAddr(), "tcp", proxier)
+	if err != nil {
+		glog.Errorf("Failed to connect to balancer: %v", err)
+		inConn.Close()
```
after you close the socket you can not try to do the copy loop.  You have to return or something

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -133,15 +133,18 @@ func (tcp *tcpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
+		inConn.Close()
 	}
+	// Spin up an async copy loop.
+	go proxyTCP(inConn.(*net.TCPConn), outConn.(*net.TCPConn))
```
why run this as another goroutine - you're already in a goroutine, you can just run proxyTCP directly.

roackb2 replied:
yes I should change this to be run directly

roackb2 committed:
change connectBalancer to connnectToBackend
close inConn after getting error
run proxyTCP directly instead of as goroutine

roackb2 commented:
@thockin  I've updated all the parts that you commented.
Thanks for your reviewing.

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -133,15 +133,19 @@ func (tcp *tcpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
-		}
-		// Spin up an async copy loop.
-		go proxyTCP(inConn.(*net.TCPConn), outConn.(*net.TCPConn))
+		go connectToBackend(service, inConn, proxier)
```
I'm sorry I didn't see this earlier - why is this done in the TCP-specific path?  The TODO that I think you're referring to is actually in the generic (TCP and UDP) path - UDP matters too. (also you should remove the TODO when you fix it :)

What about something like:

* rename `tryConnect` to `connectAndProxy`.
* pass a func(in, out net.Conn) to connectAndProxy() - one for TCP, one for UDP in their respective paths
* let connectAndProxy() call through to the function, which calls through to the proxy loop
* run connectAndProxy() as a goroutine

UDP is a little tricker, but we should solve both,

roackb2 replied:
Good advise, I'll try to implement this but may takes some time.

ncdc commented:
@kubernetes/rh-networking

thockin commented:
Closes #13176

roackb2 committed:
rename tryConnect to connectAndProxy, run as goroutine for both TCP and UDP request

roackb2 committed:
rename local variables more properly

roackb2 commented:
@thockin  sorry for bothering but I falsely pulled and rebased on Kubernetes master branch before I push to this PR, and now Google bot complains about that I'm including other people's commits into my PR. 
Should I rollback other files or any other action should I take?

thockin commented:
This shows you have 19 commits in this PR.  try this:

git fetch upstream
git rebase -i upstream/master
    delete any commits that are not yours, then save

then `git rebase -i upstream/master` again and you should see only your commit.

roackb2 commented:
@thockin  I proceed with your suggestions but I have to force push to take out other parts that I didn't modify,  and now Google bot thinks that I changed too much things and label it as size XXL ... 

In addition, I implemented the more general `connectAndProxy` function. As to the UDP path, It's working fine and I tried to make the code clear. May need your review to know how I can improve it.

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -86,17 +91,22 @@ func (tcp *tcpProxySocket) ListenPort() int {
 
-func tryConnect(service proxy.ServicePortName, srcAddr net.Addr, protocol string, proxier *Proxier) (out net.Conn, err error) {
+// Generic function type for proxying TCP and UDP request
+type proxyMethod func(in, out net.Conn, err error)
```
it's a little weird to pass errors INTO a function - can you document why and what the function is supposed to do with it?

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -104,9 +114,12 @@ func tryConnect(service proxy.ServicePortName, srcAddr net.Addr, protocol string
+		err = fmt.Errorf("failed to connect to an endpoint.")
 	}
-	return nil, fmt.Errorf("failed to connect to an endpoint.")
+	proxy(inConn, outConn, err)
```
comment on why you are calling this even with a non-nil error

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -86,17 +91,22 @@ func (tcp *tcpProxySocket) ListenPort() int {
+// Generic function type for proxying TCP and UDP request
+type proxyMethod func(in, out net.Conn, err error)
+
+func connectAndProxy(service proxy.ServicePortName, inConn net.Conn, srcAddr net.Addr, protocol string, proxier *Proxier, proxy proxyMethod) {
```
comment that this expects to be running in a goroutine

thockin commented:
Overall this looks pretty good, but it makes me super nervous.

Please fixup and rebase to head and let's get a fresh e2e run on it.

thockin commented:
Also, the code freeze for v1.1 is coming, so time is of the essence.

roackb2 commented:
If I rebase on master won't other people's commit got included?

thockin commented:
no, rebase means to take your patches off the top of your stack, sync to
head, then re-apply your patches.

git fetch upstream
git rebase -i upstream/master (you should only see your commits listed)

On Mon, Sep 14, 2015 at 11:53 PM, Jay Liang <notifications@github.com>
wrote:

> If I rebase on master won't other people's commit got included?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13423#issuecomment-140297414>
> .
>

roackb2 committed:
move error handling outside proxy method

roackb2 commented:
I moved error handling outside the proxy methods and is now located at `connectAndProxy` function, but this means we need condition checks to handle tcp or udp specific errors. 
I'm not sure whether this would be a good pattern, but otherwise I have no idea for better workflow.

I'm really thankful for your being patient. Hope everything is ok in production.

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -104,9 +115,19 @@ func tryConnect(service proxy.ServicePortName, srcAddr net.Addr, protocol string
+		err = fmt.Errorf("failed to connect to an endpoint.")
+	}
+	if err != nil {
+		if protocol == "tcp" {
```
This is a little smelly to me.  The main reason you can't just call inConn.Close() is that you're passing the UDP listener as the inConn.  I think a better pattern is something like this, tell me what you think:

in here:

```
if err != nil {
    glog.Errorf("Failed to connect to balancer: %v", err)
    inConn.Close()
    return
}
```

in UDP code:

```
// fakeUDPConn pretends to be a "connection" to a UDP client so that other code can safely call Close().
// Of course, UDP doesn't really have a connection, so just fake it.
type fakeUDPConn struct {
    net.Conn
}

func (u *fakeUDPConnection) Close() {
    // Do nothing
}
```

In UDP connectBackend:

```
go connectAndProxy(service, fakeUDPConnection{udp}, cliAddr, "udp", proxier, func(in, svrConn net.Conn) {
   // everything else
}
```

Another alternative would be to pass a closeInput function to connectAndProxy() and make TCP's be `inConn.Close()` and UDP's be an empty function.

roackb2 replied:
I think the first one is more self-documented and readable, I'll try it.

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -216,54 +230,51 @@ func (udp *udpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
 }
 
-func (udp *udpProxySocket) getBackendConn(activeClients *clientCache, cliAddr net.Addr, proxier *Proxier, service proxy.ServicePortName, timeout time.Duration) (net.Conn, error) {
+func (udp *udpProxySocket) connectBackend(activeClients *clientCache, buffer *[bufferSize]byte, n int, cliAddr net.Addr, proxier *Proxier, service proxy.ServicePortName, timeout time.Duration) {
```
it might be cleaner to just fold this func into the PorxyLoop function

roackb2 replied:
I'll use an anonymous function to wrap it and put inside the ProxyLoop function, since I think putting lots of code between lock/unlock pairs is harmful for readability.
If this is not the case or there is any other better approach please let me know.

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -216,54 +230,51 @@ func (udp *udpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
+				return
+			}
+			activeClients.clients[cliAddr.String()] = svrConn
+			go func(cliAddr net.Addr, svrConn net.Conn, activeClients *clientCache, timeout time.Duration) {
```
why not write the data first and then call proxyClient without another goroutine?

roackb2 replied:
should proxyClient still be called inside an anonymous function to utilize `defer util.HandleCrash()` to handle crash?

thockin commented:
Sorry for the delay - this is a tricky PR and requires me to page-in all the UDP knowledge I hoped never to need again :)

roackb2 committed:
use fakeUDPConn structure for safely close UDP 'connection'
fold connectBackend function into ProxyLoop function
call proxyClient directly instead of using goroutine

roackb2 commented:
@thockin  I'm getting similar error messages when testing on master branch.
Unit-test, integration-test both failed on master branch, since related to PR #14302 .
Should I do anything to fix it or just wait for the godep issue to be fixed?

roackb2 commented:
@thockin  Need I do anything for this branch to be merged?

ArtfulCoder commented:
I am still trying to convince myself that the code changes are correct. (I think there is a bug/issue, but I only reviewed it on my phone.)

I will review it again tomorrow..

roackb2 commented:
Ok, I'll review the code. If you find any problem, please let me know, thx.

thockin commented:
Jay,

Sorry.  we've just been reallllly busy on v1.1 critical changes.  I have
not forgotten this.

On Thu, Oct 15, 2015 at 7:42 PM, Jay Liang <notifications@github.com> wrote:

> @thockin <https://github.com/thockin> Need I do anything for this branch
> to be merged?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13423#issuecomment-148582746>
> .
>

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -216,61 +239,57 @@ func (udp *udpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
+			// If this is a client we know already, reuse the connection and goroutine.
+			svrConn, found := activeClients.clients[cliAddr.String()]
+			if !found {
+				// TODO: This could spin up a new goroutine to make the outbound connection,
```
shouldn't this TODO be removed now?

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -216,61 +239,57 @@ func (udp *udpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
+						return
+					}
+					activeClients.clients[cliAddr.String()] = svrConn
+					writeData(svrConn, buffer, n, timeout)
```
I think this is a problem.  buffer is in the ProxyLoop scope, and will be reused.  Since this runs in a goroutine now, it will race - the ProxyLoop might receive another packet before the old packet is written to the backend.  I think you have to block on a mutex or make a copy of the buffer.

ArtfulCoder replied:
exactly..

thockin commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -216,61 +239,57 @@ func (udp *udpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
+						glog.Errorf("SetDeadline failed: %v", err)
+						return
+					}
+					activeClients.clients[cliAddr.String()] = svrConn
```
This is not right - you're accessing activeClients in a goroutine which does not hold the mutex.  The caller might have already released the lock.

thockin commented:
The more I look at this - especially the UDP path, the more worried I am.  It's REALLY subtle code in a code path which is already sort of on its last leg (iptables proxier obsoletes this).

Do we really need to do this at this point?  I feel bad asking that, because i know this PR has taken a long time to pull together, and I like it overall -- I am just scared of it.  Curious to hear Abhi's analysis.

What if we just did the TCP path and maybe consider UDP as a separate PR?

ArtfulCoder commented:
I only looked at the udp related changes -

Correctness issues:

* had noted the buffer overwrite issue, but Tim already called it https://github.com/kubernetes/kubernetes/pull/13423#discussion_r42210291

* With this change, we are re-ordering the user-sent data itself.
   k8s udp client sends A, B.
   KubeProxy reads A, invokes routine 1
   KubeProxy reads B, invokes routine 2
   Routine 2 acquires fd lock for srvConn and sends B
   Routine 1 acquires fd lock for srvConn and sends A
   
    Ideally, udp apps are meant to handle out-of-turn data at an app-level, but, it is not an ideal world, and I would prefer that kube-proxy not aggravate out-of-turn data arrival at the final endpoint. 

Performance issues:

* the count of go-routines seems to grow unbounded (1 per incoming udp packet). It is an unknown, as to how this would perform in heavy load. Go says that go-routines are managed by the go-runtime and may or may not map to threads. In our case, I suspect they will because of the system calls to write to socket..

thockin commented:
goroutines should not be unbounded it should be 1 per Service-port and one
per client IP, modulo timeouts.

On Thu, Oct 15, 2015 at 11:39 PM, Abhi Shah <notifications@github.com>
wrote:

> I only looked at the udp related changes -
>
> Correctness issues:
>
>    -
>
>    had noted the buffer overwrite issue, but Tim already called it #13423
>    (comment)
>    <https://github.com/kubernetes/kubernetes/pull/13423#discussion_r42210291>
>    -
>
>    With this change, we are re-ordering the user-sent data itself.
>    k8s udp client sends A, B.
>    KubeProxy reads A, invokes routine 1
>    KubeProxy reads B, invokes routine 2
>    Routine 2 acquires fd lock for srvConn and sends B
>    Routine 1 acquires fd lock for srvConn and sends A
>
>    Ideally, udp apps are meant to handle out-of-turn data at an
>    app-level, but, it is not an ideal world, and I would prefer that
>    kube-proxy not aggravate out-of-turn data arrival at the final endpoint.
>
> Performance issues:
>
>    - the count of go-routines seems to grow unbounded (1 per incoming udp
>    packet). It is an unknown, as to how this would perform in heavy load. Go
>    says that go-routines are managed by the go-runtime and may or may not map
>    to threads. In our case, I suspect they will because of the system calls to
>    write to socket..
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13423#issuecomment-148629352>
> .
>

ArtfulCoder commented:
My suggestions:
For udp: parallelism should be done at a clientAddr level, with these rules -  
* data from the same active client, should be handled by the same go-routine. (waiting on a channel)
* we should never have a situation where data from the same active client is handled by multiple go-routines in parallel.
* packets from 2 different clientAddrs may go to the same go-routine.

I am worried about introducing this PR in 1.1...

thockin commented:
There's no way this can make 1.1, which is what prompts me to ask if it is
worth doing at all.  The TCP path at least looks straightforward, so i
could see doing that in 1.2 (we should benchmark this change before
merging).  UDP is scarier, and probably less impactful overall.

On Thu, Oct 15, 2015 at 11:45 PM, Abhi Shah <notifications@github.com>
wrote:

> My suggestions:
> For udp: parallelism should be done at a clientAddr level, with these
> rules -
>
>    - data from the same active client, should be handled by the same
>    go-routine. (waiting on a channel)
>    - we should never have a situation where data from the same active
>    client is handled by multiple go-routines in parallel.
>    - packets from 2 different clientAddrs may go to the same go-routine.
>
> I am worried about introducing this PR in 1.1...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13423#issuecomment-148630938>
> .
>

ArtfulCoder commented:
In the current sate, the PR seems to have no bounds on the goroutines created per service-port.

thockin commented:
go connectAndProxy() is the only goroutine called and it is only called if
the client was not previously known.  Clients that timeout should terminate
their goroutines.  Maybe I am missing something.

On Thu, Oct 15, 2015 at 11:48 PM, Abhi Shah <notifications@github.com>
wrote:

> In the current sate, the PR seems to have no bounds on the goroutines
> created per service-port.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13423#issuecomment-148631504>
> .
>

ArtfulCoder commented:
the way you read it is correct. i.e. goroutines are not unbounded, and out-of-order data should not happen either.
The inner anonymous function threw me off. have generally seem that in the context of a go routine invocation, but here the go routine is called inside the anonymous function.

thockin commented:
The other concerns stand - there are at least a couple races in this as-is.

On Fri, Oct 16, 2015 at 12:18 AM, Abhi Shah <notifications@github.com>
wrote:

> the way you read it is correct. i.e. goroutines are not unbounded, and
> out-of-order data should not happen either.
> The inner anonymous function threw me off. have generally seem that in the
> context of a go routine invocation, but here the go routine is called
> inside the anonymous function.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13423#issuecomment-148635105>
> .
>

roackb2 commented:
Thockin's concern makes sense, I know the iptables implementation is going to replace userspace implementation, so maybe we just split this to two PRs and make changes related to tcp path to be concerned first? As to the udp part, it's even ok for me to not making a PR, it's simply good to have a chance to practice. But if you maintainers think it's worthing doing, I'm also glad to cooperate. Please make a decision and I'll update this branch accordingly.

ArtfulCoder commented:
1. To improve udp performance, we could consider opening multiple kube-proxy ports per service and then fanning out the udp requests to these kube-proxy ports based on client source ip.
This would give us true udp parallelism, but requires a bigger change. We can do that for 1.2, IF, userspace proxier is still going to be around.

2. udp ends up being used mostly for dns, and we run one instance of skydns in the cluster. With this assumption, the udp changes won't help much since there will be only one dns enpoint.

So, maybe, we can leave the udp changes out for now, and we can see what the tcp changes are.

@thockin do you agree ?

thockin commented:
That was my feeling, yeah.

bgrant0607 commented:
Removing from 1.1

roackb2 commented:
Sorry for the delay, I was busying some other project. If I recover the UDP path to how it looks like in master branch, the `tryConnect` function should be kept as is. Won't this be a redundant comparing to the `connectAndProxy` function in my branch?

thockin commented:
Jay,

This PR fell off the bottom of my email.  Given the subtlety of it and the
fact that we're moving to the iptables proxier more, I am not sure we
should pursue this.  The risk outweighs the benefits.

Unless you disagree strongly, in which case we can talk numbers, I'm
closing this.

I really appreciate the effort that went into this PR - I hope we can get
you involved in other PRs.  Community will make or break this project.

On Sat, Nov 14, 2015 at 10:50 PM, Jay Liang <notifications@github.com>
wrote:

> Sorry for the delay, I was busying some other project. If I recover the
> UDP path to how it looks like in master branch, the tryConnect function
> should be kept as is. Won't this be a redundant comparing to the
> connectAndProxy function in my branch?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13423#issuecomment-156786636>
> .
>

roackb2 commented:
Thanks for all your efforts, too.  I'll try to participate another PR if I find one that I could help.
